pkg/controller/bastion/actuator.go

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud"
+	alicloudclient "github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud/client"
+	alicloudApi "github.com/gardener/gardener-extension-provider-alicloud/pkg/apis/alicloud"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/utils"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/vpc"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	"github.com/gardener/gardener/extensions/pkg/controller/bastion"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// SSHPort is the default SSH Port used for bastion ingress firewall rule
+	SSHPort = "22/22"
+)
+
+type actuator struct {
+	client            client.Client
+	logger            logr.Logger
+	alicloudECSClient alicloudclient.ECS
+	alicloudVPCClient alicloudclient.VPC
```

Comment:
Do not store clients for specific Shoot in the actuator as multiple Bastions may be reconciled at the same time. Use local clients instead.

Reply:
You can also add the client factory as a property of actuator, so that you don't need to instantiate new factories each time.

Reply:
>You can also add the client factory as a property of actuator, so that you don't need to instantiate new factories each time.

Could you point to some sample code that I can have a look thanks?   Is it looks like this?  https://github.com/gardener/gardener-extension-provider-alicloud/blob/714fc0e6d4973ef8c8ad3d9fc298b79006e8b8c6/pkg/controller/backupbucket/actuator.go#L47

Reply:
You can have a look at infrastructure controller:
https://github.com/gardener/gardener-extension-provider-alicloud/blob/master/pkg/controller/infrastructure/actuator.go#L99-L100

Note that the ECS client here for infrastructure is Seed's client, and the client per shoot is created locally in each reconciliation flow.

Reply:
committed in b1de2d2b